src/main/java/com/snowflake/kafka/connector/internal/SnowflakeInternalStage.java

```diff
@@ -12,50 +14,89 @@
 import net.snowflake.client.jdbc.SnowflakeFileTransferAgent;
 import net.snowflake.client.jdbc.SnowflakeFileTransferConfig;
 import net.snowflake.client.jdbc.SnowflakeFileTransferMetadataV1;
+import net.snowflake.client.jdbc.SnowflakeSQLException;
 import net.snowflake.client.jdbc.cloud.storage.StageInfo;
+import net.snowflake.client.jdbc.internal.apache.commons.io.FilenameUtils;
 
+/**
+ * Implementation of put API through JDBC's API uploadWithoutConnection.
+ *
+ * <p>We fetch the credentials and cache it for AWS and Azure. We will refresh if cache hits 30 mins
+ * (Cache Eviction)
+ *
+ * <p>For GCS, we dont have any cache, we will make a call to GS for every put API since we require
+ * presignedURL
+ */
 public class SnowflakeInternalStage extends Logging {
 
   private static class SnowflakeMetadataWithExpiration {
-    SnowflakeFileTransferMetadataV1 fileTransferMetadata;
-    long timestamp;
+
+    /* File transfer metadata when fetched from GS call */
+    private final SnowflakeFileTransferMetadataV1 fileTransferMetadata;
+
+    /* Timestamp when GS call happened */
+    private final long timestampLastCredentialFetchMillis;
+
+    /* StageType, can be fetched from fileTransferMetadata */
+    private final StageInfo.StageType stageType;
 
     SnowflakeMetadataWithExpiration(
         SnowflakeFileTransferMetadataV1 fileTransferMetadata, long timestamp) {
       this.fileTransferMetadata = fileTransferMetadata;
-      this.timestamp = timestamp;
+      this.timestampLastCredentialFetchMillis = timestamp;
+      this.stageType = this.fileTransferMetadata.getStageInfo().getStageType();
+    }
+
+    public StageInfo.StageType getStageType() {
+      return this.stageType;
     }
   }
 
   // Any operation on the map should be atomic
   private final ConcurrentMap<String, SnowflakeMetadataWithExpiration> storageInfoCache =
       new ConcurrentHashMap<>();
 
-  public static String dummyPutCommandTemplate =
-      "PUT file:///tmp/dummy_location_kakfa_connector_tmp/ @";
+  // GCS Put version requires the dummy command to have filename and entire filePath including
+  // stageName after "@"
+  // For example: PUT file:///fileName @stageName/app/table/partition
+  public static String dummyPutCommandTemplateGCS = "PUT file:///%s @%s/%s";
```

Comment:
Template being different is very tricky and nuanced. Can we add a little bit more details in comments for future readers of the code?

Reply:
Will add more details. (I have an example already but I see how it can be confusing)

Reply:
https://snowflakecomputing.atlassian.net/browse/SNOW-350676